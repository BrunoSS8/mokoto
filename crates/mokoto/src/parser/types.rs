use super::*;

fn opt_mutability(p: &mut Parser) {
    let c = p.checkpoint();
    if p.eat(VarKw) {
        p.finish_at(c, MutModifier)
    }
}

fn opt_typ_args(p: &mut Parser) {
    let c = p.checkpoint();
    if p.eat(Lt) {
        typ(p);
        while p.eat(Comma) {
            typ(p);
        }
        if !p.eat(Gt) {
            // TODO Error
        }
        p.finish_at(c, TypArgs)
    }
}

fn path_or_name(p: &mut Parser) -> Option<Checkpoint> {
    assert!(p.at(Ident));
    let c = p.checkpoint();
    p.bump(Ident);
    if !p.at(Dot) {
        return Some(c);
    }
    while p.eat(Dot) {
        // TODO: Error
        p.bump(Ident);
    }
    p.finish_at(c, Path);
    None
}

fn path(p: &mut Parser) {
    path_or_name(p).map(|c| p.finish_at(c, Path));
}

fn array_typ(p: &mut Parser) {
    assert!(p.at(LBracket));
    let c = p.checkpoint();
    p.bump(LBracket);
    opt_mutability(p);
    typ(p);
    p.bump(RBracket);
    p.finish_at(c, ArrayT)
}

fn paren_or_tuple_typ(p: &mut Parser) {
    assert!(p.at(LParen));
    let c = p.checkpoint();
    p.bump(LParen);
    if p.eat(RParen) {
        p.finish_at(c, TupT);
        return;
    }
    typ_item(p);
    if p.eat(RParen) {
        p.finish_at(c, ParenT);
        return;
    }
    while p.eat(Comma) {
        typ_item(p);
    }
    // TODO Error
    p.bump(RParen);
    p.finish_at(c, TupT)
}

fn typ_item(p: &mut Parser) {
    let c = p.checkpoint();
    if p.at(Ident) {
        match path_or_name(p) {
            Some(c_label) if p.at(Colon) => {
                p.finish_at(c_label, Name);
                p.bump(Colon);
                typ(p);
                p.finish_at(c, NamedT);
            }
            Some(c_path) => {
                p.finish_at(c_path, Path);
                opt_typ_args(p);
                p.finish_at(c, PathT);
                if p.eat(Arrow) {
                    typ(p);
                    p.finish_at(c, FuncT);
                }

            }
            None => {
                opt_typ_args(p);
                p.finish_at(c, PathT);
            }
        };
    } else {
        typ(p);
    }
}

fn typ_nullary(p: &mut Parser) {
    match p.peek() {
        LParen => paren_or_tuple_typ(p),
        LBracket => array_typ(p),
        Ident => {
            let c = p.checkpoint();
            path(p);
            opt_typ_args(p);
            p.finish_at(c, PathT)
        }
        _ => unreachable!("What"),
    }
}

fn typ_un(p: &mut Parser) {
    let c = p.checkpoint();
    if p.eat(Question) {
        typ_un(p);
        p.finish_at(c, OptionalT);
    } else {
        typ_nullary(p)
    }
}

pub(super) fn typ(p: &mut Parser) {
    let c = p.checkpoint();
    typ_un(p);
    if p.eat(Arrow) {
        typ(p);
        p.finish_at(c, FuncT);
    }
}
