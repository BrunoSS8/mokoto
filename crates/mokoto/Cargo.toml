[package]
name = "mokoto"
version = "0.1.0"
authors = ["Christoph Hegemann <christoph.hegemann1337@gmail.com>"]
edition = "2018"
license = "MIT OR Apache-2.0"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
logos = "0.11.4"
num-derive = "0.3.3"
num-traits = "0.2.14"
rowan = "0.10.0"

[dev-dependencies]
insta = { version = "1", features = ["glob"] }
