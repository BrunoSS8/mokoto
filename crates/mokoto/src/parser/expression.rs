mod atom;
use super::*;

pub(super) fn expr(p: &mut Parser) {
    if atom::literal(p) {
        return;
    }
    unreachable!("Can only do literals so far.")
}
