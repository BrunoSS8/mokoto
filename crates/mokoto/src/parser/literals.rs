use super::{Parser, TokenSet};
use crate::lexer::SyntaxKind::*;

// TODO: All literals
pub(crate) const STARTS_LIT: TokenSet = TokenSet::new(&[NULL_KW, TRUE_KW, FALSE_KW, NUMBER_LIT]);

pub fn literal(p: &mut Parser) {
    let c = p.checkpoint();
    if p.at_ts(STARTS_LIT) {
        p.bump_any();
        p.finish_at(c, LITERAL)
    } else {
        // TODO: Error
        unreachable!()
    }
}
