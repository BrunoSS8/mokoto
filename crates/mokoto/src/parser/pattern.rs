use super::Parser;
use crate::lexer::SyntaxKind;

pub(super) fn pat(p: &mut Parser) {
    match p.peek() {
        SyntaxKind::Ident => var_pattern(p),
        _ => unreachable!("Unknown pattern."),
    }
}

fn var_pattern(p: &mut Parser) {
    p.start_node(SyntaxKind::VarP);
    p.bump(SyntaxKind::Ident);
    p.finish_node();
}
