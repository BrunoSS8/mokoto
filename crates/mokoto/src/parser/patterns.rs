use super::literals::{literal, STARTS_LIT};
use super::types::opt_annot;
use super::Parser;
use crate::lexer::SyntaxKind::*;

pub(super) fn pattern(p: &mut Parser) {
    pat_un(p)
}

fn paren_or_tuple_pattern(p: &mut Parser) {
    assert!(p.at(L_PAREN));
    let c = p.checkpoint();
    p.bump(L_PAREN);
    if p.eat(R_PAREN) {
        p.finish_at(c, TUPLE_PAT);
        return;
    }
    pattern(p);
    if p.eat(R_PAREN) {
        p.finish_at(c, PAREN_PAT);
        return;
    }
    while p.eat(COMMA) {
        pattern(p);
    }
    // TODO Error
    p.bump(R_PAREN);
    p.finish_at(c, TUPLE_PAT)
}

fn pat_field(p: &mut Parser) -> bool {
    let c = p.checkpoint();
    if !p.eat(IDENT) {
        return false;
    }
    p.finish_at(c, NAME);
    opt_annot(p);
    if p.eat(EQUALS) {
        pattern(p);
        p.finish_at(c, PATTERN_FIELD_PAT)
    } else {
        p.finish_at(c, PATTERN_FIELD_PUN)
    }
    true
}

fn pat_un(p: &mut Parser) {
    let c = p.checkpoint();
    match p.current() {
        HASH => {
            p.bump(HASH);
            if !p.eat(IDENT) {
                // TODO: Error
                unreachable!()
            }
            p.finish_at(c, TAG);
            let _ = pat_nullary(p);
            p.finish_at(c, VARIANT_PAT)
        }
        _ => {
            if !pat_nullary(p) {
                // TODO: Error
                unreachable!()
            }
        }
    }
}

fn pat_nullary(p: &mut Parser) -> bool {
    if p.at(L_BRACE) {
        let c = p.checkpoint();
        p.bump(L_BRACE);
        while pat_field(p) {
            if !p.eat(SEMICOLON) {
                break;
            }
        }
        if !p.eat(R_BRACE) {
            unreachable!()
        }
        p.finish_at(c, OBJECT_PAT);
        true
    } else {
        pat_plain(p)
    }
}

fn pat_plain(p: &mut Parser) -> bool {
    let c = p.checkpoint();
    match p.current() {
        UNDERSCORE => {
            p.bump(UNDERSCORE);
            p.finish_at(c, WILDCARD_PAT);
        }
        IDENT => {
            p.bump(IDENT);
            p.finish_at(c, NAME);
            p.finish_at(c, VAR_PAT)
        }
        L_PAREN => paren_or_tuple_pattern(p),
        t if STARTS_LIT.contains(t) => {
            literal(p);
            p.finish_at(c, LITERAL_PAT)
        }
        // TODO: Error
        _ => return false,
    };
    true
}
