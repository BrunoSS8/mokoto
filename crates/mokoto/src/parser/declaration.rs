use super::expression;
use super::pattern;
use super::Parser;
use crate::lexer::SyntaxKind;

pub(super) fn decl(p: &mut Parser) {
    match p.current() {
        SyntaxKind::LetKw => let_decl(p),
        _ => unreachable!("Unknown decl."),
    }
}

fn ident(p: &mut Parser) {
    if !p.eat(SyntaxKind::Ident) {
        p.error("Expected an ident")
    }
}

fn let_decl(p: &mut Parser) {
    p.start_node(SyntaxKind::LetD);
    p.bump(SyntaxKind::LetKw);
    pattern::pat(p);
    if !p.eat(SyntaxKind::Equals) {
        p.error("Expected an =");
    }
    expression::expr(p);
    p.finish_node();
}
