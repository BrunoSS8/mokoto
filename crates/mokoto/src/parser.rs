mod declaration;
mod expression;
mod pattern;
mod token_set;
mod types;

use crate::lexer::{Lexer, SyntaxKind, SyntaxKind::*, Token};
use crate::syntax::{MotokoLanguage, SyntaxNode};
use declaration::decl;
use rowan::{Checkpoint, GreenNode, GreenNodeBuilder, Language};
use token_set::TokenSet;

pub struct Parser<'a> {
    tokens: Vec<Token<'a>>,
    builder: GreenNodeBuilder<'static>,
    // TODO: Be smarter here
    errors: Vec<String>,
}

impl<'a> Parser<'a> {
    pub fn new(input: &'a str) -> Self {
        let mut tkns = Lexer::lex(input);
        tkns.reverse();
        Self {
            tokens: tkns,
            builder: GreenNodeBuilder::new(),
            errors: vec![],
        }
    }

    pub fn parse(mut self) -> Parse {
        self.start_node(SyntaxKind::Root);

        decl(&mut self);

        self.finish_node();

        Parse {
            green_node: self.builder.finish(),
            errors: self.errors,
        }
    }

    pub fn parse_typ(mut self) -> Parse {
        self.start_node(SyntaxKind::Root);

        types::typ(&mut self);

        self.finish_node();

        Parse {
            green_node: self.builder.finish(),
            errors: self.errors,
        }
    }

    fn start_node(&mut self, kind: SyntaxKind) {
        self.builder.start_node(MotokoLanguage::kind_to_raw(kind));
    }

    fn start_node_at(&mut self, checkpoint: Checkpoint, kind: SyntaxKind) {
        self.builder
            .start_node_at(checkpoint, MotokoLanguage::kind_to_raw(kind));
    }

    fn finish_node(&mut self) {
        self.builder.finish_node();
    }

    fn finish_at(&mut self, c: Checkpoint, kind: SyntaxKind) {
        self.start_node_at(c, kind);
        self.finish_node();
    }

    fn checkpoint(&self) -> Checkpoint {
        self.builder.checkpoint()
    }

    fn bump_any(&mut self) {
        let (leading, (kind, text), trailing) = self.tokens.pop().unwrap();

        for (kind, text) in leading {
            self.builder
                .token(MotokoLanguage::kind_to_raw(kind), text.into());
        }

        self.builder
            .token(MotokoLanguage::kind_to_raw(kind), text.into());

        for (kind, text) in trailing {
            self.builder
                .token(MotokoLanguage::kind_to_raw(kind), text.into());
        }
    }

    fn bump(&mut self, kind: SyntaxKind) {
        assert!(self.eat(kind))
    }

    fn eat(&mut self, kind: SyntaxKind) -> bool {
        if self.current() != kind {
            return false;
        }
        self.bump_any();
        true
    }

    fn nth(&mut self, n: usize) -> SyntaxKind {
        let len = self.tokens.len();
        if n >= len {
            SyntaxKind::Eof
        } else {
            self.tokens[n - len].1 .0
        }
    }

    fn nth_at(&mut self, n: usize, kind: SyntaxKind) -> bool {
        self.nth(n) == kind
    }

    fn current(&mut self) -> SyntaxKind {
        self.nth(0)
    }

    fn at(&mut self, kind: SyntaxKind) -> bool {
        self.nth_at(0, kind)
    }

    fn at_ts(&mut self, kinds: TokenSet) -> bool {
        kinds.contains(self.current())
    }

    fn error(&mut self, msg: &str) {
        self.errors.push(msg.to_string())
    }
}

pub struct Parse {
    green_node: GreenNode,
    errors: Vec<String>,
}

impl Parse {
    pub fn debug_tree(&self) -> String {
        let syntax_node = SyntaxNode::new_root(self.green_node.clone());
        let formatted = format!("{:#?}", syntax_node);

        // We cut off the last byte because formatting the SyntaxNode adds on a newline at the end.
        formatted[0..formatted.len() - 1].to_string()
    }
}
