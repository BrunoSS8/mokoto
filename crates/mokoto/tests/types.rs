use mokoto::parser::{Parser};

use insta::{assert_snapshot, glob};

#[test]
fn parse_types() {
    use std::fs;

    glob!("passing/types/*.mo", |path| {
        let input = fs::read_to_string(path).unwrap();
        for inp in input.split("---\n") {
            let parse = Parser::new(&inp).parse_typ();
            let output = format!("{}\n---\n{}", inp, parse.debug_tree());
            assert_snapshot!(output);
        }
    });
}
