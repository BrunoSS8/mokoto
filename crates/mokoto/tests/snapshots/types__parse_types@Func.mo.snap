---
source: crates/mokoto/tests/types.rs
expression: output
input_file: crates/mokoto/tests/passing/types/Func.mo
---
A -> B

---
Root@0..6
  FuncT@0..6
    PathT@0..2
      Path@0..2
        Ident@0..1 "A"
        Space@1..2 " "
    Arrow@2..4 "->"
    Space@4..5 " "
    PathT@5..6
      Path@5..6
        Ident@5..6 "B"
